packages/block-library/src/media-text/deprecated.js

```diff
@@ -143,6 +205,154 @@ const v4ToV5Supports = {
+		__experimentalFontWeight: true,
+		__experimentalFontStyle: true,
+		__experimentalTextTransform: true,
+		__experimentalTextDecoration: true,
+		__experimentalLetterSpacing: true,
+		__experimentalDefaultControls: {
+			fontSize: true,
+		},
+	},
+};
+
+// Version with wide as the default alignment.
+// See: https://github.com/WordPress/gutenberg/pull/46971
+const v6 = {
+	attributes: v6Attributes,
+	supports: v6Supports,
+	save( { attributes } ) {
+		const {
+			isStackedOnMobile,
+			mediaAlt,
+			mediaPosition,
+			mediaType,
+			mediaUrl,
+			mediaWidth,
+			mediaId,
+			verticalAlignment,
+			imageFill,
+			focalPoint,
+			linkClass,
+			href,
+			linkTarget,
+			rel,
+		} = attributes;
+		const mediaSizeSlug =
+			attributes.mediaSizeSlug || DEFAULT_MEDIA_SIZE_SLUG;
+		const newRel = isEmpty( rel ) ? undefined : rel;
+
+		const imageClasses = classnames( {
+			[ `wp-image-${ mediaId }` ]: mediaId && mediaType === 'image',
+			[ `size-${ mediaSizeSlug }` ]: mediaId && mediaType === 'image',
+		} );
+
+		let image = (
+			<img
+				src={ mediaUrl }
+				alt={ mediaAlt }
+				className={ imageClasses || null }
+			/>
+		);
+
+		if ( href ) {
+			image = (
+				<a
+					className={ linkClass }
+					href={ href }
+					target={ linkTarget }
+					rel={ newRel }
+				>
+					{ image }
+				</a>
+			);
+		}
+
+		const mediaTypeRenders = {
+			image: () => image,
+			video: () => <video controls src={ mediaUrl } />,
+		};
+		const className = classnames( {
+			'has-media-on-the-right': 'right' === mediaPosition,
+			'is-stacked-on-mobile': isStackedOnMobile,
+			[ `is-vertically-aligned-${ verticalAlignment }` ]:
+				verticalAlignment,
+			'is-image-fill': imageFill,
+		} );
+		const backgroundStyles = imageFill
+			? v6ImageFillStyles( mediaUrl, focalPoint )
+			: {};
+
+		let gridTemplateColumns;
+		if ( mediaWidth !== DEFAULT_MEDIA_WIDTH ) {
+			gridTemplateColumns =
+				'right' === mediaPosition
+					? `auto ${ mediaWidth }%`
+					: `${ mediaWidth }% auto`;
+		}
+		const style = {
+			gridTemplateColumns,
+		};
+
+		if ( 'right' === mediaPosition ) {
+			return (
+				<div { ...useBlockProps.save( { className, style } ) }>
+					<div
+						{ ...useInnerBlocksProps.save( {
+							className: 'wp-block-media-text__content',
+						} ) }
+					/>
+					<figure
+						className="wp-block-media-text__media"
+						style={ backgroundStyles }
+					>
+						{ ( mediaTypeRenders[ mediaType ] || noop )() }
+					</figure>
+				</div>
+			);
+		}
+		return (
+			<div { ...useBlockProps.save( { className, style } ) }>
+				<figure
+					className="wp-block-media-text__media"
+					style={ backgroundStyles }
+				>
+					{ ( mediaTypeRenders[ mediaType ] || noop )() }
+				</figure>
+				<div
+					{ ...useInnerBlocksProps.save( {
+						className: 'wp-block-media-text__content',
+					} ) }
+				/>
+			</div>
+		);
+	},
+	migrate: migrateDefaultAlign,
+	isEligible( attributes ) {
+		return attributes.align === undefined;
+	},
```

Comment:
One part that I don't understand here - wow do we avoid this running for newly added blocks that default to 'none'?

I think they would also have `attributes.align === undefined` given default values aren't serialized.

Reply:
Thanks for flagging that. 

It looks like the piece of the puzzle I was not expecting was that the custom class name block support catches the old `alignwide` class instead of the block's markup becoming invalid. 

I think this means we need either:
1. To specify an `isEligible` callback with the ability to check both the `align` attribute and any existing `alignwide` class.
2. To be able to make the `alignwide` class in the block markup be invalid i.e. not added to the `className` attribute via the custom class names block support.

The catch so far with option `#1` is that when the deprecation's `isEligible` check is running, the custom class name block support hasn't yet updated the `className` attribute. 

As for option `#2`, I've only had a quick play so far tweaking where the [custom class names are determined](https://github.com/WordPress/gutenberg/blob/05381d2074f6cfd9d1b6d0f779943291203b829d/packages/blocks/src/api/parser/fix-custom-classname.js#L44) but haven't had any success there yet.

I'll need to revisit this one later when the mind is a bit fresher I think. 

Any ideas or suggestions on alternative solutions are more than welcome 🙂

Reply:
FWIW, if we passed the final block attributes through to the `isEligible` functions we'd have what we need to correctly determine if the deprecation should run.

With the diff below, all the current fixtures pass and you can see in the editor either by adding current blocks and reloading or pasting into the code editor, that the latest version doesn't get re-migrated.

<details>
<summary>Example (hacky) diff</summary>

```diff
diff --git a/packages/block-library/src/media-text/deprecated.js b/packages/block-library/src/media-text/deprecated.js
index e5a0f6980d..68765edb65 100644
--- a/packages/block-library/src/media-text/deprecated.js
+++ b/packages/block-library/src/media-text/deprecated.js
@@ -346,8 +346,11 @@ const v6 = {
 		);
 	},
 	migrate: migrateDefaultAlign,
-	isEligible( attributes ) {
-		return attributes.align === undefined;
+	isEligible( attributes, innerBlocks, finalAttributes ) {
+		return (
+			attributes.align === undefined &&
+			!! finalAttributes.className?.includes( 'alignwide' )
+		);
 	},
 };
 
diff --git a/packages/blocks/src/api/parser/apply-block-deprecated-versions.js b/packages/blocks/src/api/parser/apply-block-deprecated-versions.js
index 2a3e880a3f..7679252d53 100644
--- a/packages/blocks/src/api/parser/apply-block-deprecated-versions.js
+++ b/packages/blocks/src/api/parser/apply-block-deprecated-versions.js
@@ -50,7 +50,11 @@ export function applyBlockDeprecatedVersions( block, rawBlock, blockType ) {
 		const { isEligible = stubFalse } = deprecatedDefinitions[ i ];
 		if (
 			block.isValid &&
-			! isEligible( parsedAttributes, block.innerBlocks )
+			! isEligible(
+				parsedAttributes,
+				block.innerBlocks,
+				block.attributes
+			)
 		) {
 			continue;
 		}

```
</details>

Reply:
Now https://github.com/WordPress/gutenberg/pull/48815 has landed, we can leverage `isEligible` to prevent re-migrating blocks that don't need to be. 

See: https://github.com/WordPress/gutenberg/pull/48404/commits/295ca16f3edc26a871263f0373de62b2609948e4